### PR TITLE
Add support for Ephemeral Containers to the kubelet

### PR DESCRIPTION
**Pull request by verb**
**What this PR does / why we need it**: This implements the [Troubleshooting Running Pods](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/troubleshoot-running-pods.md) in the kubelet, protected by a feature flag.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
WIP #10834
WIP #27140

**Special notes for your reviewer**:
/assign @yujuhong 

**Release note**:

```release-note
Ephemeral containers have been added in alpha. These temporary containers can be added to running pods for purposes such as debugging, similar to how `kubectl exec` runs an process in an existing container. Also like `kubectl exec`, no resources are reserved for ephemeral containers and they are not restarted when they exit.

Note that container namespace targeting is not yet implemented, so [process namespace sharing](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) must be enabled to view process from other containers in the pod.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-node/20190212-ephemeral-containers.md
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -354,9 +355,10 @@ type containerToKillInfo struct {
 	message string
 }
 
-// podActions keeps information what to do for a pod.
+// podActions keeps information what to do for a pod based on its spec.
+// Note that podActions will not affect Debug Containers since they are not part of the spec.
 type podActions struct {
-	// Stop all running (regular and init) containers and the sandbox for the pod.
+	// Stop all running (regular, init and debug) containers and the sandbox for the pod.
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: This seems to conflict with the above comment stating that the actions don't affect debug containers.

File: pkg/kubelet/kuberuntime/labels.go [hunk 2]
```diff
@@ -105,6 +106,14 @@ func newContainerLabels(container *v1.Container, pod *v1.Pod, containerType kube
 	labels[types.KubernetesContainerNameLabel] = container.Name
 	if utilfeature.DefaultFeatureGate.Enabled(features.DebugContainers) {
 		labels[types.KubernetesContainerTypeLabel] = string(containerType)
+		// It might be useful to store the spec for every container, but for now only store spec for debug containers
+		if containerType == kubecontainer.ContainerTypeEphemeral {
+			if containerJSON, e := json.Marshal(container); e != nil {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/labels.go]: Does size matter here? If it does a base64 encoded protobuf will probably be more efficient.

File: pkg/kubelet/kuberuntime/labels.go [hunk 3]
```diff
@@ -105,6 +106,14 @@ func newContainerLabels(container *v1.Container, pod *v1.Pod, containerType kube
...
+				glog.Errorf("Unable to marshal %v container %q of pod %v: %v", containerType, container, pod.Name, e)
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/labels.go]: nit: use [`format.Pod(pod)`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/format/pod.go#L32) in place of the pod name.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 4]
```diff
@@ -809,6 +834,30 @@ func (m *kubeGenericRuntimeManager) RunInContainer(id kubecontainer.ContainerID,
...
 }
 
+// StartEphemeralContainer creates a Debug Container in a pod. It's called from the same goroutine as SyncPod()
+// TODO(verb) switch v1.Container to v1.EphemeralContainer when it is added to the API
+func (m *kubeGenericRuntimeManager) StartEphemeralContainer(pod *v1.Pod, podStatus *kubecontainer.PodStatus, container *v1.Container, pullSecrets []v1.Secret) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DebugContainers) {
+		return errors.New("Debug Containers feature disabled")
+	}
+
+	if kubecontainer.GetContainerSpec(pod, container.Name) != nil {
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: What if there's another ephemeral container with the same name? Is there a race condition here?

**Comment by verb** [comment]
I've updated this PR to reflect updates to the API from kubernetes/community#1269, which makes the kubelet change simpler.

Note that tests will fail until this PR is rebased to include #59416.

**Comment by spiffxp** [comment]
ping, needs rebase

**Comment by spiffxp** [comment]
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/277 which is currently planned for v1.13

**Comment by guineveresaenger** [comment]
@verb given that this PR still is a WIP, what is your level of confidence this can be ready for the 1.13 milestone? Enhancement freeze is tomorrow end of day. If there is no communication or update on the PR, this is going to be pulled from the milestone as it doesn't fit with our "stability" theme. If there is no communication after COB tomorrow, an exception will be required to add it back to the milestone. Please let me or Enhancements Lead @kacole2 know where we stand. Thanks!

**Comment by verb** [comment]
@guineveresaenger I have the kubelet code ready, it's just waiting on the API #59416. Let's wait for the API discussion to conclude and remove this from 1.13.

**Comment by spiffxp** [comment]
/milestone v1.14
Punting out of v1.13 per discussion above

**Comment by soggiest** [comment]
Hello! We have begun Code Freeze for 1.15. Is this PR still planned to be included in 1.15? If sop will is be finished in the next week or so?

**Comment by verb** [comment]
@soggiest No, this is blocked on API review #59416, which I don't expect to be complete in time for 1.15.

**Comment by verb** [comment]
Added release notes

File: pkg/kubelet/container/helpers_test.go [hunk 5]
```diff
@@ -320,6 +324,72 @@ func TestExpandVolumeMountsWithSubpath(t *testing.T) {
...
+						{EphemeralContainerCommon: v1.EphemeralContainerCommon{
+							Name: "debug-container",
+						}},
+					},
+				},
+			},
+			haveName:      "debug-container",
+			wantContainer: &v1.Container{Name: "debug-container"},
+		},
+	} {
```
Comment (yujuhong) [inline pkg/kubelet/container/helpers_test.go]: Use subtests, i.e., `t.Run()`

File: pkg/kubelet/container/helpers_test.go [hunk 6]
```diff
@@ -320,6 +324,72 @@ func TestExpandVolumeMountsWithSubpath(t *testing.T) {
...
+		gotContainer := GetContainerSpec(tc.havePod, tc.haveName)
+		if diff := cmp.Diff(tc.wantContainer, gotContainer); diff != "" {
+			t.Errorf("GetContainerSpec for %q returned diff (-want +got):%v", tc.name, diff)
```
Comment (yujuhong) [inline pkg/kubelet/container/helpers_test.go]: Can probably just use `assert` here.

File: pkg/kubelet/kubelet_pods.go [hunk 7]
```diff
@@ -1414,13 +1414,30 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
 		len(pod.Spec.InitContainers) > 0,
 		true,
 	)
+	if utilfeature.DefaultFeatureGate.Enabled(features.EphemeralContainers) {
+		var ecSpecs []v1.Container
+		for i := range pod.Spec.EphemeralContainers {
+			ecSpecs = append(ecSpecs, v1.Container(pod.Spec.EphemeralContainers[i].EphemeralContainerCommon))
+		}
+
+		// TODO(verb): By now we've iterated podStatus 3 times. We could refactor this to make a single
```
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: File an issue instead.

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go [hunk 8]
```diff
@@ -396,17 +397,16 @@ func mountedReadOnlyByPod(podVolume v1.Volume, pod *v1.Pod) bool {
...
-	for _, container := range pod.Spec.InitContainers {
-		if !mountedReadOnlyByContainer(podVolume.Name, &container) {
-			return false
-		}
-	}
-	for _, container := range pod.Spec.Containers {
-		if !mountedReadOnlyByContainer(podVolume.Name, &container) {
+
+	mountedReadOnly := true
+	podutil.VisitContainers(&pod.Spec, func(c *v1.Container) bool {
```
Comment (yujuhong) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: cc volumemanager approvers @msau42 @saad-ali @jsafrane

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 9]
```diff
@@ -610,6 +627,7 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 //  4. Create sandbox if necessary.
 //  5. Create init containers.
 //  6. Create normal containers.
+//  7. Create ephemeral containers.
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: This seems to contradict the comment you had above " Ephemeral containers may be started even if initialization is not yet complete."

If we want to allow the ephemeral containers to be created even when init containers haven't finished running, shouldn't we create them before (5)?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 10]
```diff
@@ -739,22 +757,40 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 		return
 	}
 
-	// Step 5: start the init container.
-	if container := podContainerChanges.NextInitContainerToStart; container != nil {
-		// Start the next init container.
+	// Helper containing boilerplate common to starting all types of containers.
+	// typeName is a label used to describe this type of container in log messages.
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: nit: list the types that are currently in use.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 11]
```diff
@@ -764,29 +800,14 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
...
-		startContainerResult := kubecontainer.NewSyncResult(kubecontainer.StartContainer, container.Name)
-		result.AddSyncResult(startContainerResult)
-
-		isInBackOff, msg, err := m.doBackOff(pod, container, podStatus, backOff)
-		if isInBackOff {
-			startContainerResult.Fail(err, msg)
-			klog.V(4).Infof("Backing Off restarting container %+v in pod %v", container, format.Pod(pod))
-			continue
-		}
+		start("container", &pod.Spec.Containers[idx])
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Why discarding the returned error?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 10 above)
Comment (verb) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Done.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 11 above)
Comment (verb) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: That's the current behavior: start the container, log any errors, move on to the next container. The logging is now done in `start()`

I should mention that the only behavior here that I (intentionally) changed was that the log spam filtering for `ErrImagePullBackOff` was previously only applied to regular containers and not init containers.

File: pkg/kubelet/container/helpers_test.go (hunk 5 above)
Comment (verb) [reply pkg/kubelet/container/helpers_test.go]: Done.

File: pkg/kubelet/container/helpers_test.go (hunk 6 above)
Comment (verb) [reply pkg/kubelet/container/helpers_test.go]: I don't feel strongly, but the output of the manual diff is more readable:
```
        helpers_test.go:390: GetContainerSpec for "ephemeral container" returned diff (-want +got):  &v1.Container{
            -   Name:    "debug-containr",
            +   Name:    "debug-container",
                Image:   "",
                Command: nil,
                ... // 18 identical fields
              }
```

with assert:
```
        helpers_test.go:388:
                Error Trace:    helpers_test.go:388
                Error:          Not equal:
                                expected: &v1.Container{Name:"debug-containr", Image:"", Command:[]string(nil), Args:[]string(nil), WorkingDir:"", Ports:[]v1.ContainerPort(nil),
EnvFrom:[]v1.EnvFromSource(nil), Env:[]v1.EnvVar(nil), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList(nil)}, VolumeMounts:[]v1.VolumeMoun
t(nil), VolumeDevices:[]v1.VolumeDevice(nil), LivenessProbe:(*v1.Probe)(nil), ReadinessProbe:(*v1.Probe)(nil), Lifecycle:(*v1.Lifecycle)(nil), TerminationMessagePath:"", Terminat
ionMessagePolicy:"", ImagePullPolicy:"", SecurityContext:(*v1.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}
                                actual  : &v1.Container{Name:"debug-container", Image:"", Command:[]string(nil), Args:[]string(nil), WorkingDir:"", Ports:[]v1.ContainerPort(nil),
 EnvFrom:[]v1.EnvFromSource(nil), Env:[]v1.EnvVar(nil), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList(nil)}, VolumeMounts:[]v1.VolumeMou
nt(nil), VolumeDevices:[]v1.VolumeDevice(nil), LivenessProbe:(*v1.Probe)(nil), ReadinessProbe:(*v1.Probe)(nil), Lifecycle:(*v1.Lifecycle)(nil), TerminationMessagePath:"", TerminationMessagePolicy:"", ImagePullPolicy:"", SecurityContext:(*v1.SecurityContext)(nil), Stdin:false, StdinOnce:false, TTY:false}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,2 +1,2 @@
                                -(*v1.Container)(&Container{Name:debug-containr,Image:,Command:[],Args:[],WorkingDir:,Ports:[],Env:[],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[],LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[],TerminationMessagePolicy:,VolumeDevices:[],})
                                +(*v1.Container)(&Container{Name:debug-container,Image:,Command:[],Args:[],WorkingDir:,Ports:[],Env:[],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},},VolumeMounts:[],LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:nil,Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[],TerminationMessagePolicy:,VolumeDevices:[],})

                Test:           TestGetContainerSpec/ephemeral_container
```

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go (hunk 8 above)
Comment (msau42) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: Can this be simplified to just `return podutil.VisitContainers(...)`?

**Review by msau42 — COMMENTED** [review]
Just one minor comment for volumemanager changes.

/approve

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go (hunk 8 above)
Comment (verb) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go]: I was worried that conflating the two levels of bool returns for "visiting was completed" and "mounted read only by pod" would be confusing, but I tried it out and it makes sense for this "search for a counterexample" sort of pattern.

I've updated to this, thanks for the suggestion.

**Comment by verb** [comment]
Cleaning up labels from before api change merged.

/remove-kind api-change
/remove-area test
/remove-area release-eng
/remove-sig apps
/remove-sig auth
/remove-sig api-machinery
/remove-sig release
/remove-sig storage
/remove-sig testing

**Pull request merged**